File: src/XerahS.Mobile.Swift/XerahSMobile/Core/Data/Upload/CustomUploader.swift
//
//  CustomUploader.swift
//  XerahS Mobile (Swift)
//
//  Copyright (c) 2007-2026 ShareX Team.
//

import Foundation

/// Upload a file via custom .sxcu-style config: POST to RequestURL with multipart.
final class CustomUploader {
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 60
        config.timeoutIntervalForResource = 120
        return URLSession(configuration: config)
    }()

    func uploadFile(filePath: String, entry: CustomUploaderEntry) -> UploadOutcome {
        if entry.requestUrl.isEmpty { return .failure(error: "Request URL is empty") }
        guard FileManager.default.fileExists(atPath: filePath) else { return .failure(error: "File not found") }
        guard let url = URL(string: entry.requestUrl) else { return .failure(error: "Invalid URL") }
        let fileUrl = URL(fileURLWithPath: filePath)
        let formName = entry.fileFormName.isEmpty ? "file" : entry.fileFormName

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        for (k, v) in entry.headers where !v.isEmpty { request.setValue(v, forHTTPHeaderField: k) }

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"\(formName)\"; filename=\"\(fileUrl.lastPathComponent)\"\r\n\r\n".data(using: .utf8)!)
        if let fileData = try? Data(contentsOf: fileUrl) { body.append(fileData) }
        body.append("\r\n".data(using: .utf8)!)
        if !entry.body.isEmpty {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"body\"\r\n\r\n".data(using: .utf8)!)
            body.append(entry.body.data(using: .utf8)!)
            body.append("\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        var outcome: UploadOutcome?
        let sem = DispatchSemaphore(value: 0)
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                outcome = .failure(error: error.localizedDescription)
                sem.signal()
                return
            }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            let bodyStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
            if code < 200 || code >= 300 {
                outcome = .failure(error: "HTTP \(code): \(bodyStr.prefix(200))")
                sem.signal()
                return
            }
            let extracted = self.extractUrl(from: bodyStr, expression: entry.urlExpression)
            let urlResult = extracted ?? bodyStr.trimmingCharacters(in: .whitespacesAndNewlines).prefix(500).description
            outcome = urlResult.isEmpty ? .failure(error: "No URL in response") : .success(url: urlResult)
            sem.signal()
        }
        task.resume()
        sem.wait()
        return outcome ?? .failure(error: "Upload failed")
    }

    private func extractUrl(from responseBody: String, expression: String) -> String? {
        guard !expression.isEmpty else { return nil }
        guard let regex = try? NSRegularExpression(pattern: expression) else { return nil }
        let range = NSRange(responseBody.startIndex..., in: responseBody)
        guard let match = regex.firstMatch(in: responseBody, options: [], range: range) else { return nil }
        if match.numberOfRanges > 1, let r = Range(match.range(at: 1), in: responseBody) {
            return String(responseBody[r])
        }
        if let r = Range(match.range, in: responseBody) { return String(responseBody[r]) }
        return nil
    }
}

File: src/XerahS.Mobile.Kt/core/data/src/main/java/com/getsharex/xerahs/mobile/core/data/upload/CustomUploader.kt
/*
 * XerahS - The Avalonia UI implementation of ShareX
 * Copyright (c) 2007-2026 ShareX Team
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Optionally you can also view the license at <http://www.gnu.org/licenses/>.
 */

package com.getsharex.xerahs.mobile.core.data.upload

import com.getsharex.xerahs.mobile.core.domain.CustomUploaderEntry
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Upload a file via custom .sxcu-style config: POST to RequestURL with multipart or body.
 */
class CustomUploader(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()
) {

    fun uploadFile(filePath: String, entry: CustomUploaderEntry): UploadOutcome {
        if (entry.requestUrl.isBlank()) return UploadOutcome.Failure("Request URL is empty")
        val file = File(filePath)
        if (!file.exists()) return UploadOutcome.Failure("File not found")
        return try {
            val request = buildRequest(entry, file)
            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                return UploadOutcome.Failure("HTTP ${response.code}: ${response.body?.string()?.take(200) ?: ""}")
            }
            val bodyStr = response.body?.string() ?: ""
            val url = extractUrl(bodyStr, entry.urlExpression)
                ?: bodyStr.trim().take(500)
            if (url.isNotBlank()) UploadOutcome.Success(url) else UploadOutcome.Failure("No URL in response")
        } catch (e: Exception) {
            UploadOutcome.Failure(e.message ?: "Upload failed")
        }
    }

    private fun buildRequest(entry: CustomUploaderEntry, file: File): Request {
        val formName = entry.fileFormName.ifBlank { "file" }
        val multipart = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(formName, file.name, file.asRequestBody(null))
        if (entry.body.isNotBlank()) multipart.addFormDataPart("body", entry.body)
        val body = multipart.build()
        val builder = Request.Builder().url(entry.requestUrl).post(body)
        entry.headers.forEach { (k, v) -> if (v != null) builder.addHeader(k, v) }
        return builder.build()
    }

    private fun extractUrl(responseBody: String, expression: String): String? {
        if (expression.isBlank()) return null
        return try {
            val regex = expression.toRegex()
            regex.find(responseBody)?.groupValues?.getOrNull(1) ?: regex.find(responseBody)?.value
        } catch (e: Exception) {
            null
        }
    }
}
